flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterTableSchemaUtil.java

```diff
@@ -0,0 +1,783 @@
+                                    getComment(column));
+
+            columns.put(name, newColumn);
+        }
+
+        private void convertComputedColumn(
+                SqlTableColumn.SqlComputedColumn column, AlterSchemaStrategy strategy) {
+            String columnName = column.getName().getSimple();
+
+            Map<String, RelDataType> accessibleFieldNamesToTypes = new HashMap<>();
+            accessibleFieldNamesToTypes.putAll(physicalFieldNamesToTypes);
+            accessibleFieldNamesToTypes.putAll(metadataFieldNamesToTypes);
+
+            try {
+                final SqlNode validatedExpr =
+                        sqlValidator.validateParameterizedExpression(
+                                column.getExpr(), accessibleFieldNamesToTypes);
+                String escapedExpr = escapeExpressions.apply(validatedExpr);
+                Schema.UnresolvedColumn newColumn =
+                        columnByExpression(columnName, escapedExpr, getComment(column));
+                computedFieldNamesToTypes.put(
+                        columnName, sqlValidator.getValidatedNodeType(validatedExpr));
+                computedFieldNamesToExpressions.put(columnName, escapedExpr);
+                columns.put(columnName, newColumn);
+            } catch (Exception e) {
+                String exMsg =
+                        String.format(
+                                "Could not %s computed column `%s` AS %s.\n%s",
+                                strategy == AlterSchemaStrategy.ADD ? "create" : "modify",
+                                columnName,
+                                escapeExpressions.apply(column.getExpr()),
+                                e.getMessage());
+                if (e instanceof CalciteContextException
+                        && e.getMessage().contains("Unknown identifier")) {
+                    exMsg =
+                            String.format(
+                                    "%s\nComputed column can only be derived from existed physical or metadata column.",
+                                    exMsg);
+                }
+                throw new ValidationException(exMsg);
+            }
+        }
+
+        private void verifyComputedColumn() {
+            Map<String, RelDataType> accessibleFieldNamesToTypes = new HashMap<>();
+            accessibleFieldNamesToTypes.putAll(physicalFieldNamesToTypes);
+            accessibleFieldNamesToTypes.putAll(metadataFieldNamesToTypes);
+            for (Map.Entry<String, String> entry : computedFieldNamesToExpressions.entrySet()) {
+                CalciteParser parser = new CalciteParser(parserConfig);
+                String columnName = entry.getKey();
+                SqlNode expr = parser.parseExpression(entry.getValue());
+                try {
+                    sqlValidator.validateParameterizedExpression(expr, accessibleFieldNamesToTypes);
+                } catch (Throwable e) {
```

Comment:
The reason why catch `Throwable` instead of `Exception` (or, to be specific, `CalciteContextException`) is because `sqlValidator#validateParameterizedExpression` might throw 
- NPE, E.g. `e ROW<f0 STRING, f1 INT, f2 ROW<DOUBLE, ARRAY<FLOAT>>>, f AS e.f1 + e.f2.f0`, and `ALTER TABLE MODIFY e STRING`
- `AssertionError` when `inferReturnType`